gslib/wildcard_iterator.py

```diff
@@ -180,10 +182,16 @@ def __iter__(self, bucket_listing_fields=None,
         else:
           yield bucket_listing_ref
       else:
-        # By default, assume a non-wildcarded URL is an object, not a prefix.
-        # This prevents unnecessary listings (which are slower, more expensive,
-        # and also subject to eventual consistency).
-        if (not ContainsWildcard(self.wildcard_url.url_string) and
+        # By default, assume a non-wildcarded and non-prefix ending URL is an object.
+        # The prefix-ending check case is important since it prevents us from reporting
+        # a prefix as an object in the output instead of listing the contents, as the user
+        # probably intended for us to when an object and prefix with the same root name
+        # exist in the bucket. i.e: gs:bucket//foo/ and gs:bucket//foo
+        # Checking object(s) first prevents unnecessary listings (which are slower,
+        # more expensive, and also subject to eventual consistency).
+        contains_wildcard = ContainsWildcard(self.wildcard_url.url_string)
+        is_prefix = IsCloudSubdirPlaceholder(self.wildcard_url_raw)
```

Comment:
It looks like `IsCloudSubdirPlaceholder` is used here for a side effect, which is just to figure out if the is a prefix. By using this function you end up roping in objects with names like `folder_$folder$` and I don't think that's what you want. I would just replace this call with `self.wildcard_url_raw.endswith('/')`.

Reply:
Done.